arkfilepicker/src/main/java/space/taran/arkfilepicker/presentation/filepicker/ArkFilePickerFragment.kt

```diff
@@ -0,0 +1,429 @@
+                        resources.getColor(
+                            R.color.ark_file_picker_gray,
+                            null
+                        )
+                    )
+                    tv.setOnClickListener {
+                        viewModel.onItemClick(fullPathToPart)
+                    }
+                }
+                layoutPath.addView(tv)
+            }
+        scrollPath.post {
+            scrollPath.fullScroll(ScrollView.FOCUS_RIGHT)
+        }
+    }
+
+    override fun getTheme() = themeId!!
+
+    private fun initBackButtonListener() {
+        requireDialog().setOnKeyListener { _, keyCode, keyEvent ->
+            if (keyCode == KeyEvent.KEYCODE_BACK &&
+                keyEvent.action == KeyEvent.ACTION_UP
+            ) {
+                viewModel.onBackClick()
+            }
+            return@setOnKeyListener true
+        }
+    }
+
+    companion object {
+        const val PATH_PICKED_REQUEST_KEY = "arkFilePickerPathPicked"
+        const val PATH_PICKED_PATH_BUNDLE_KEY = "arkFilePickerPathPickedPathKey"
+        const val FOLDER_CHANGED_REQUEST_KEY = "arkFilePickerFolderChanged"
+        const val FOLDER_CHANGED_FOLDER_BUNDLE_KEY =
+            "arkFilePickerFolderChangedFolderKey"
+
+        fun newInstance(config: ArkFilePickerConfig) =
+            ArkFilePickerFragment().apply {
+                titleStringId = config.titleStringId
+                pickButtonStringId = config.pickButtonStringId
+                cancelButtonStringId = config.cancelButtonStringId
+                internalStorageStringId = config.internalStorageStringId
+                accessDeniedStringId = config.accessDeniedStringId
+                itemsPluralId = config.itemsPluralId
+                themeId = config.themeId
+                initialPath = config.initialPath?.toString()
+                showRoots = config.showRoots
+                pathPickedRequestKey = config.pathPickedRequestKey
+                mode = config.mode.ordinal
+            }
+
+        private const val DIALOG_WIDTH = 300f
+        private const val PATH_PART_PADDING = 4f
+    }
+}
+
+internal class FilesPage(
+    private val fragment: Fragment,
+    private val viewModel: ArkFilePickerViewModel,
+    private val itemsPluralId: Int
+) : AbstractBindingItem<ArkFilePickerItemFilesRootsPageBinding>() {
+    private val filesAdapter = ItemAdapter<FileItem>()
+    private var currentFiles = emptyList<Path>()
+
+    override val type = 0
+
+    override fun createBinding(
+        inflater: LayoutInflater,
+        parent: ViewGroup?
+    ) = ArkFilePickerItemFilesRootsPageBinding.inflate(inflater, parent, false)
+
+    override fun bindView(
+        binding: ArkFilePickerItemFilesRootsPageBinding,
+        payloads: List<Any>
+    ) = with(binding) {
+        rvFiles.adapter = FastAdapter.with(filesAdapter)
+        viewModel.observe(fragment, ::render)
+    }
+
+    private fun render(state: FilePickerState) {
```

Comment:
Maybe we could use generic types for state? This would allow us to separate one state type with optional fields into `FilesTabState` and `RootsTabState`.

Reply:
I'm using a framework that allows only one state per viewmodel. Yes, we can add `FilesTabState` and `RootsTabState` as fields. But now we only need a list of files, is it worth creating a separate state for this?

Reply:
I just don't really like that we mix 2 states into one, could be dirty state eventually:
```
data class FilePickerState(
    val devices: List<Path>,                    // only for RootsPage?
    val selectedDevicePos: Int,                 // only for RootsPage
    val currentPath: Path,                      // only for FilesPage?
    val files: List<Path>,                      // only for FilesPage
    val rootsWithFavs: Map<Path, List<Path>>    // only for RootsPage
)
```